library/serialize.c

```diff
@@ -0,0 +1,333 @@
+ *   0x45 ('E') execute   target->host  Execute a function
+ *   0x50 ('P') push      target->host  Push an input for the next execution
+ *   0x72 ('r') result    host->target  Return an output from the last execution
+ *
+ * The host maintains a stack of parameters and results. Stack
+ * elements are byte arrays. This stack is initially empty. The "push"
+ * command transmits one item from the target to the host and pushes
+ * this item to the stack. The "execute" command requests that the host
+ * perform an action, taking parameters from the stack and sending outputs
+ * back to the target.
+ *
+ * Format of a push packet (by byte offset):
+ *   0..0       0x50    Type identifier
+ *   1..3       size    Number of bytes to follow
+ *   4..3+size  data    Content of the byte array
+ *
+ * Format of an execute packet (by byte offset):
+ *   0..0       0x45    Type identifier
+ *   1..3       func    Function identifier
+ * Function identifiers are conventionally arranged as follows:
+ *   bits 23-16   Library code (1=sockets)
+ *   bits 15-8    Function identifier inside the library
+ *   bits 7-4     Number of parameters
+ *   bits 3-0     Number of results after status code on success
+ *
+ * Format of a result packet (by byte offset):
+ *   0..0       0x72    Type identifier
+ *   1..3       size    Number of bytes to follow
+ *   4..3+size  data    Content of the byte array
+ *
+ * All functions currently follow the following result status
+ * convention. The first result is a status code (MBEDTLS_ERR_xxx
+ * constant). If this is 0 then more results follow. If this is
+ * nonzero then it is the only result.
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include <limits.h>
+#include <stdint.h>
+#include <string.h>
+#include <signal.h>
+#include "mbedtls/serialize.h"
+
+#if defined(MBEDTLS_SERIALIZE_C)
+
+static int serialize_write_fd = -1;
+static int serialize_read_fd = -1;
+
+#if defined(MBEDTLS_SERIALIZE_FORK_FRONTEND_C)
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <signal.h>
+
+static int host_pid;
+
+static void kill_host_frontend( void )
+{
+    kill( host_pid, SIGHUP );
+}
+
+static void relay_signal( int signum )
+{
+    /* Let the atexit function kill the frontend */
+    exit( signum + 128 );
+}
+
+#define CHECK( expr, var )                      \
+    do {                                        \
+        var = ( expr );                         \
+        if( var < 0 )                           \
+        {                                       \
+            perror( #expr );                    \
+            goto cleanup;                       \
+        }                                       \
+    } while( 0 );
+static int mbedtls_serialize_prepare( void )
+{
+    int ret;
+    int null = -1;
+    int host_to_target[2] = {-1, -1}, target_to_host[2] = {-1, -1};
```

Comment:
I'd prefer `null_fd` as the variable name here; it hints at the meaning and prevents any accidental confusion with the NULL pointer.